runtime/j9vm/jvm.c

```diff
@@ -1218,73 +1226,148 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
 			goto bail;
 		}
 
-		version = props_file_get(classlibProps, "version");
+		version = props_file_get(propsFile, "version");
 		if (NULL == version) {
-	#ifdef DEBUG
-		printf("No 'version' property in %s\n", classlibProps);
-	#endif
+#ifdef DEBUG
+			printf("No 'version' property in %s\n", propsFile);
+#endif
 			goto bail;
 		}
 	
 		decoded = decodeSetting("shape", shape, SHAPE_SETTINGS, NUM_SHAPE_SETTINGS);
-		if (decoded == 0) {
+		if (0 == decoded) {
 			goto bail;
 		}
 		finalVersion |= decoded;
 	
 		decoded = decodeSetting("version", version, VERSION_SETTINGS, NUM_VERSION_SETTINGS);
-		if (decoded == 0) {
+		if (0 == decoded) {
 			goto bail;
 		}
 		finalVersion |= decoded;
 
-	bail:
-		props_file_close(classlibProps);
+bail:
+		props_file_close(propsFile);
+	} else {
+#ifdef DEBUG
+		printf("Could not open %s\n", propsFile);
+#endif
+	}
+	
+	return finalVersion;
+}
+
+/**
+ * Get Java version info from 'release' file
+ * Attempt loading 'release' file, if found, 'JAVA_VERSION' value is retrieved, and decoded as following:
+ * "1.8.0_xxx"	--- Java 8 (J2SE_18 | J2SE_SHAPE_SUN)
+ * "9"			--- Java 9 (J2SE_19 | J2SE_SHAPE_B165)
+ * "10"			--- Java 18.3 (J2SE_B7 | J2SE_SHAPE_B1803)
+ * Others		--- (ßJ2SE_LATEST | J2SE_SHAPE_LATEST)
+ * Note: 'release' file contains JAVA_VERSION="10" for Java 18.3 at this moment.
+ *
+ * @return jvm SE Version value
+ */
+static UDATA
+getVersionFromReleaseFile(void)
+{
+	PORT_ACCESS_FROM_PORT(&j9portLibrary);
+	J9StringBuffer *propsPathBuffer = NULL;
+	j9props_file_t propsFile = NULL;
+	UDATA finalVersion = 0;
+
+	propsPathBuffer = jvmBufferCat(propsPathBuffer, jvmBufferData(j9Buffer));
+	propsPathBuffer = jvmBufferCat(propsPathBuffer, DIR_SEPARATOR_STR "release");
+	propsFile = props_file_open(PORTLIB, jvmBufferData(propsPathBuffer), NULL, 0);
+	free(propsPathBuffer);
+	propsPathBuffer = NULL;
+	if (NULL != propsFile) {
+		const char *version = props_file_get(propsFile, "JAVA_VERSION");
+		if (NULL != version) {
+#define		JAVA_VERSION_8		"\"1.8.0"		/* its usual format is "1.8.0_xxx" */
+			if (!strncmp(version, JAVA_VERSION_8, strlen(JAVA_VERSION_8))) {
+#undef		JAVA_VERSION_8
+				finalVersion = J2SE_18 | J2SE_SHAPE_SUN;
+			} else if (!strcmp(version, "\"9\"")) {
+				finalVersion = J2SE_19 | J2SE_SHAPE_B165;
+			} else if (!strcmp(version, "\"10\"")) {
+				finalVersion = J2SE_B7 | J2SE_SHAPE_B1803;
+			} else {
+				/* Assume latest Java version and shape */
+				finalVersion = J2SE_LATEST | J2SE_SHAPE_LATEST;
+			}
+		} else {
+#ifdef DEBUG
+			printf("No 'JAVA_VERSION' property in %s\n", propsFile);
+#endif
+		}
+		props_file_close(propsFile);
+	} else {
+#ifdef DEBUG
+		printf("Could not open %s\n", propsFile);
+#endif
+	}
+
+	return finalVersion;
+}
+
+/**
+ * Get Java version of running JVM
+ * Attempt loading the 'classlib.properties' file first, otherwise try 'release' file,
+ * if neither of files is found, J2SE_LATEST | J2SE_SHAPE_LATEST will be returned.
+ *
+ * @return jvmSEVersion value
```

Comment:
Please update the @return as per the comments on getVersionFromClasslibPropertiesFile(). jvmSEVersion is an internal implementation detail.